### ISSUE TITLE
bad log_prob values from AutoregressiveRationalQuadraticSpline

### ISSUE THREAD

**Issue by Clemente90** [VincentStimper/normalizing-flows#24]
Hello, first I want to say that this is a really cool library, great job!

I was learning about and experimenting with normalizing flows using this library starting with the example notebooks that are provided. I made a couple of modifications and began seeing log likelihood values that do not seem correct.

I defined my flow model like so:

```
def splineFlowModelConstructor(base_dist, latent_size, hidden_units_per_flow_block = 256, hidden_layers_per_flow_block = 2, num_layers_in_model = 32):
  # Define list of flows
  flows = []
  for i in range(num_layers_in_model):
      # Neural network with two hidden layers having 64 units each
      # Last layer is initialized by zeros making training more stable
      flows += [nf.flows.AutoregressiveRationalQuadraticSpline(latent_size, hidden_layers_per_flow_block, hidden_units_per_flow_block)]
      flows += [nf.flows.LULinearPermute(latent_size)]

  # Construct flow model
  model = nf.NormalizingFlow(base_dist, flows)
  return model
```

And my base and target distributions:

```
N = 8
base = nf.distributions.base.GaussianMixture(1, N, loc=torch.zeros(1, N), scale=1.0 * torch.ones(1, N))
target = nf.distributions.base.GaussianMixture(4, N, loc=40.0 * torch.rand(4, N))
```

Looking at the first 4 of 8 dimensions, it looks like the NF model fits the data quite well:
![image](https://user-images.githubusercontent.com/12226343/226470782-75d80e24-b6b6-4637-aa99-9dfcba6d2530.png)

And in the reverse direction from the target distribution samples through the model and plotted against the base distribution:
![image](https://user-images.githubusercontent.com/12226343/226470561-e5bb58d5-d134-4f24-b3cf-ceaa600ef246.png)

The problem is when I run model.log_prob( samples ) I get really large values:

```
model.eval()
model.cpu()

sample = target.sample(100).float()
sample -= mean
sample /= stdev

probs = model.log_prob( sample )

print(probs)
print(torch.exp(probs)[0].item())
```

Output:

```
tensor([ 5.4699,  6.4249,  7.8194,  6.3206,  7.5243,  5.1152,  5.9383,  5.5442,
         6.2432,  5.4947,  6.8944, -0.6838,  4.9015,  3.7826,  2.4365,  2.1861,
         4.3309,  6.7556,  6.9447,  5.0362,  7.6152,  7.5655,  6.9386,  6.1994,
         3.9858,  5.7908,  5.1715,  8.4266,  7.3475,  6.3807,  6.9416,  5.8353,
         4.7075,  5.3876,  7.7150,  8.0491,  4.9048,  6.5957,  6.0490,  7.3908,
         4.8446,  6.9842,  3.0278,  5.0358,  7.4219,  8.0341,  6.4992,  2.5209,
         0.4988,  3.3599,  3.3678,  6.3835,  3.1170,  4.1072,  6.1465,  5.8682,
        -0.0433,  4.1949,  8.7846, -1.4585,  4.4729,  7.9508,  0.6114,  4.1969,
         5.7270,  6.0191,  5.7380,  5.4258,  3.0765,  1.0877,  5.6288,  7.1016,
         3.5127,  3.6730,  1.7773,  6.5651,  5.0568,  2.3242,  4.6668,  4.2033,
         6.3310,  4.6139,  6.9517,  7.2128,  4.2161,  8.1131,  5.9864,  5.3085,
         3.9264,  4.5639,  5.2759,  8.6771,  4.5639,  4.9707,  6.1224,  6.1103,
         3.8134,  3.1536,  4.1925,  0.8635], grad_fn=<AddBackward0>)
237.43670654296875
```

Those are supposed to be negative log_prob values, right? And 237.4 is supposed to be a number between [0, 1], right?

Best,

**Comment by Clemente90** [comment]
Would running target.log_prob() on the samples prior to the re-scaling work the same way?

```
model.eval()
model.cpu()

sample = target.sample(100).float()
sample_scaled = sample - m.cpu()
sample_scaled /= s.cpu()

probs_from_model = model.log_prob(sample_scaled)
probs_from_target_dist = target.log_prob(sample)

print(probs_from_model)
print(torch.exp(probs_from_model)[0].item())

print()

print(probs_from_target_dist)
print(torch.exp(probs_from_target_dist)[0].item())
```

Output:
```
tensor([ 4.4903,  4.2997,  5.4324,  3.9816,  4.4711,  3.5592,  4.3241,  4.6952,
         1.5439,  4.6124,  4.3070,  3.1918,  3.4568,  6.2067,  5.7150,  3.2984,
         0.4182,  4.1373,  7.2445,  4.8905,  4.5713,  1.5819,  2.7022,  5.3710,
         3.2661,  6.3267,  3.0108,  1.2476,  1.6435,  3.1418,  4.8486,  3.3067,
         2.0900,  1.6301,  2.7087,  3.3445,  2.0032,  6.2780,  6.0869,  6.2271,
         5.4194,  3.9321,  5.5718,  4.4330,  3.2285,  2.5771,  6.5517,  4.1202,
         4.9087,  3.3947,  5.0558,  6.1823,  6.1326, -1.1190,  2.5255,  4.3414,
         4.8667,  1.7161,  4.2921,  6.9498,  3.5424,  5.3509,  2.8694,  0.6522,
         2.2838,  4.5669,  5.8214,  4.3999, -0.5175,  5.4123,  5.3807, -0.1436,
         4.7292,  1.7416,  7.1322,  6.3167,  4.2389,  6.7894,  3.9714,  3.8585,
         1.5911, -0.2556,  4.1990,  4.4847,  5.2921, -1.8543,  3.3106,  3.8633,
         3.1255,  6.4033,  0.2822,  7.3791,  4.7763,  7.2801,  3.9587,  3.5378,
         1.6269,  1.5234,  1.9431,  5.3408], grad_fn=<AddBackward0>)
89.15181732177734

tensor([-11.9530, -11.5577, -10.9878, -12.8910, -11.4824, -12.3896, -11.9103,
        -13.3553, -14.8448, -11.4620, -10.9395, -13.0726, -13.3437, -10.2955,
        -12.0687, -13.7796, -16.2041, -11.3506,  -9.6583, -12.4364, -11.9083,
        -15.7696, -13.7087, -12.0690, -12.9692, -10.2296, -15.5404, -15.6063,
        -15.6523, -14.5685, -12.3384, -12.4044, -13.1801, -13.4737, -12.7037,
        -12.7887, -14.0471, -10.2147, -10.6740, -11.6499, -11.3376, -12.9450,
        -10.2181, -12.4265, -15.3826, -14.8013, -10.1166, -13.2219, -12.4427,
        -12.9142, -10.9406, -10.7000, -11.7691, -19.9174, -13.8676, -14.2307,
        -11.9494, -15.3586, -12.6462, -10.0540, -12.0239, -10.2631, -13.9995,
        -15.7287, -12.8565, -13.1267, -11.9490, -11.4505, -18.1301, -11.1184,
        -12.0171, -17.6683, -12.5626, -14.6238, -10.8127, -10.8127, -12.4227,
        -12.4115, -13.3735, -13.4995, -13.1689, -16.1271, -14.6353, -11.2491,
        -10.4271, -15.5246, -13.2947, -11.2839, -12.3496, -10.4752, -14.2652,
        -10.7427, -11.1497,  -9.4200, -13.4084, -12.5513, -16.1202, -15.2385,
        -13.4398, -10.6935], dtype=torch.float64, grad_fn=<LogsumexpBackward0>)
6.439657747876614e-06
```

**Comment by Clemente90** [comment]
Thank you, I appreciate your responses.

The one thing I am not understanding is, ```exp(0.0) == 1.0```. Any log-probability value that is greater than ```0.0``` would imply a probability density function value greater than ```1.0```.  So the distribution could not integrate to ```1.0```. (Assuming the target distribution is not a delta function.)

**Comment by VincentStimper** [comment]
Hi @Clemente90,

this is not true. In the first method to compute the integral I mentioned above, you would multiply with the volume of the bin, which will be small, pushing the value of the term in the sum below 1. To do importance sampling, you would compute the mean, so some of the terms can be larger than 1 while the result could still be 1. Feel free to implement them to convince yourself that the distribution is indeed normalized.

Best regards,
Vincent

**Comment by Clemente90** [comment]
I am sorry, you are right. It makes complete sense for those log_prob values to be greater than 0.0. That is very embarrassing to make such a basic conceptual mistake. Thank you for your patient explanations.

Best,

**Comment by VincentStimper** [comment]
Hi @Clemente90,

no worries, this is indeed counter-intuitive at first glance.

I'm glad I could help. I'm closing the issue now.

Best regards,
Vincent
